### DISCUSSION TITLE
Simple example for mapping attributes

### DISCUSSION THREAD

**Discussion by TutanRamon** [CuyZ/Valinor#279] [Q&A] [answered]
First, thanks for this package. I know I need it because i don't want to have all kind of coding lines with mappings in my Controller. But, i don't know how to start.

I have a incoming post request from an Excel file. Unfortunatly, I can't change how the post is made by VBA. The names of the attributes are not equal to the attribute names of my Model. 

So, I receive a JSON post which I am trying to map to my model, using Valinor. 

For the sake of simplicity, lets say I receive this JSON :

```
[
  {
    "Company Name": "Abbott Laboratories",
    "Ticker": "ABT",
    "Transactions": [
    {
      "Date": "2015-05-30T22:00:00.000Z",
      "Total Amount of shares": 1000,
      "Currency": "USD"
    },
    {
      "Date": "2022-02-20T22:00:00.000Z",
      "Total Amount of shares": 500,
      "Currency": "USD"
    }
    ]
  },
  {
    "Company Name": "ASML Holding N.V.",
    "Ticker": "ASML",
    "Transactions": [
    {
      "Date": "2020-04-30T22:00:00.000Z",
      "Total Amount of shares": 100,
      "Currency": "EUR"
    },
    {
      "Date": "2021-01-23T22:00:00.000Z",
      "Total Amount of shares": 200,
      "Currency": "EUR"
    }
    ]
  }
]
```

Now, my model is called "Stock" and has 2 attributes: "name" and "ticker".  I have another model called StockTransactions which has a 1:n relation with Stock (one stock can have many transactions). 

I created a StockData class:

```
final class StockData
{
    public function __construct(
        /** @var non-empty-string */
        public readonly string $Naam,
        public readonly string $Ticker
    ) {
    }
}

```

In my api controller I have this method:

```
public function store(Request $request)
    {
        try {
            $stock = (new \CuyZ\Valinor\MapperBuilder())
                ->mapper()
                ->map(StockData::class, \CuyZ\Valinor\Mapper\Source\Source::json(json_encode($request->json()->all())));

            echo $stock->name
        } catch (\CuyZ\Valinor\Mapper\MappingError $error) {
            // Handle the error…
        }
   }
```

Nothing is happening. I read something about mapping the attributes, but I want the mapping to be done in the StockData class and not in the controller. How can I achieve this?

**Comment by romm** [comment, marked as answer]
Hi @TutanRamon, I have several suggestion to make:

It seems that the request arguments describe a list of `StockData`, so it looks like you need to map not to the class directly, rather a list of it:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->mapper()
    ->map('list<' . StockData::class . '>', /* … */);
```

Note that it would also work with an array. You can find the list of available types [in the documentation](https://valinor.cuyz.io/latest/usage/type-reference/).

Next, the request object seems to be able to provide an array of raw data, you probably don't need to use the `JsonSource`. You probably can do something like the example below, as long as you provide an iterable to the mapper.

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->mapper()
    ->map('list<' . StockData::class . '>', $request->all());
```

Because the source contains superfluous keys (more data than actually needed), you need to activate an option in the mapper builder:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->allowSuperfluousKeys()
    ->mapper()
    ->map('list<' . StockData::class . '>', $request->all());
```

The keys in the input do not match the properties of your DTO, so you need to tell the mapper about it. This is done by using a custom source like below. More information can be found [in the documentation](https://valinor.cuyz.io/latest/how-to/transform-input/).


```php
$source = $source = Source::iterable($request->all())->map([
    '*.Company Name' => 'Naam'
]);

(new \CuyZ\Valinor\MapperBuilder())
    ->allowSuperfluousKeys()
    ->mapper()
    ->map('list<' . StockData::class . '>', $source);
```

Finally, you obviously need to handle the error. I can't be of any help because it will really depend on how you can/need to manage it in your application, but you can find more information about this topic [in the documentation](https://valinor.cuyz.io/latest/usage/error-handling/).

Hope this helps.

**Reply by TutanRamon** [reply]
> Hi @TutanRamon, I have several suggestion to make:
>  ...
> Hope this helps.


This certainly helps. Thanks a lot for providing this answer. The reason I have a StockData.class is because I thought that was necessary. I might have an update method as well in my controller and I don't want to have the mapping definitions in both methods. So, I thought: i create a data class for my Stock model and call it StockData. But, for now it looks like the only thing that class should do, is return an array of mappings. Is that right?

**Comment by TutanRamon** [comment]
Ok, I am a bit further. However, when I look at the JSON that is being sent, I see basiscally a double nested array.

Each "stock" holds an array called "Transactions". Within this array, there is an array of Transaction.

The JSON looks like this:
```
{
    "Company Name": "Abbott Laboratories",
    "Ticker": "ABT",
    "Transactions": [
    {
      "Date": "2015-05-30T22:00:00.000Z",
      "Total Amount of shares": 1000,
      "Currency": "USD"
    },
    {
      "Date": "2022-02-20T22:00:00.000Z",
      "Total Amount of shares": 500,
      "Currency": "USD"
    }
    ]
  },
```

I am trying to get to the level of transactions. I got it working, but the output is not what I want it to be.

```
final class StockData
{
    public readonly string $name;
    public readonly string $ticker;

    /** @var list<Transactions> */
    public readonly array $transactions;
}

final class Transactions
{
    /** @var string[] */
    public readonly array $transaction;
}

final class Transaction
{
    public readonly string $currency;
    public readonly int $totalAmountOfShares;
 
}
```

My Controller:

```
$source = \CuyZ\Valinor\Mapper\Source\Source::array($request->json()->all())->map([
    '*.Company Name' => 'name',
    '*.Ticker' => 'ticker',
    '*.Transactions' => 'transactions',
    '*.Transactions.*.*.Currency' => 'currency',
    '*.Transactions.*.*.Total Amount of shares' => 'totalAmountOfShares',
]);
```

**Reply by romm** [reply]
You do not need to map all properties, you can use the `camelCaseKeys()` method:

```php
$source = \CuyZ\Valinor\Mapper\Source\Source::array($request->json()->all())->map([
    '*.Company Name' => 'name',
])->camelCaseKeys();
```

For the transactions, there is no need for two classes, one is enough:

```php
final class StockData
{
    public readonly string $name;
    public readonly string $ticker;

    /** @var list<Transaction> */
    public readonly array $transactions;
}

final class Transaction
{
    public readonly string $currency;
    public readonly int $totalAmountOfShares;
 
}
```
